### PR TITLE
Forced Bucketing - Implementation 

### PR DESCRIPTION
**Pull request by oakbani**
✏️ Forced Bucketing feature
✏️ Unit test cases

### CODE REVIEW

**Comment by mikeng13** [comment]
build

**Comment by mikeng13** [comment]
build

**Comment by mikeng13** [comment]
@oakbani I am getting the following errors from running our e2e tests. Seems like a legitimate crash:

```
2017-09-28 16:35:28 - NoMethodError - undefined method `present?' for nil:NilClass:
	/usr/local/bundle/bundler/gems/ruby-sdk-ce610654b93e/lib/optimizely/project_config.rb:283:in `get_forced_variation'
	/usr/local/bundle/bundler/gems/ruby-sdk-ce610654b93e/lib/optimizely/decision_service.rb:63:in `get_variation'
	/usr/local/bundle/bundler/gems/ruby-sdk-ce610654b93e/lib/optimizely.rb:139:in `get_variation'
	/usr/local/bundle/bundler/gems/ruby-sdk-ce610654b93e/lib/optimizely.rb:104:in `activate'
	./main.rb:33:in `block in <main>'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1632:in `call'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1632:in `block in compile!'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:991:in `block (3 levels) in route!'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1010:in `route_eval'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:991:in `block (2 levels) in route!'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1037:in `block in process_route'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1035:in `catch'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1035:in `process_route'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:989:in `block in route!'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:988:in `each'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:988:in `route!'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1094:in `block in dispatch!'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `block in invoke'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `catch'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `invoke'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1091:in `dispatch!'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:923:in `block in call!'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `block in invoke'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `catch'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1073:in `invoke'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:923:in `call!'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:913:in `call'
	/usr/local/bundle/gems/rack-protection-2.0.0/lib/rack/protection/xss_header.rb:18:in `call'
	/usr/local/bundle/gems/rack-protection-2.0.0/lib/rack/protection/path_traversal.rb:16:in `call'
	/usr/local/bundle/gems/rack-protection-2.0.0/lib/rack/protection/json_csrf.rb:26:in `call'
	/usr/local/bundle/gems/rack-protection-2.0.0/lib/rack/protection/base.rb:50:in `call'
	/usr/local/bundle/gems/rack-protection-2.0.0/lib/rack/protection/base.rb:50:in `call'
	/usr/local/bundle/gems/rack-protection-2.0.0/lib/rack/protection/frame_options.rb:31:in `call'
	/usr/local/bundle/gems/rack-2.0.3/lib/rack/logger.rb:15:in `call'
	/usr/local/bundle/gems/rack-2.0.3/lib/rack/common_logger.rb:33:in `call'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:231:in `call'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:224:in `call'
	/usr/local/bundle/gems/rack-2.0.3/lib/rack/head.rb:12:in `call'
	/usr/local/bundle/gems/rack-2.0.3/lib/rack/method_override.rb:22:in `call'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/show_exceptions.rb:22:in `call'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:194:in `call'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1955:in `call'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1499:in `block in call'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1726:in `synchronize'
	/usr/local/bundle/gems/sinatra-2.0.0/lib/sinatra/base.rb:1499:in `call'
	/usr/local/bundle/gems/rack-2.0.3/lib/rack/handler/webrick.rb:86:in `service'
	/usr/local/lib/ruby/2.3.0/webrick/httpserver.rb:140:in `service'
	/usr/local/lib/ruby/2.3.0/webrick/httpserver.rb:96:in `run'
	/usr/local/lib/ruby/2.3.0/webrick/server.rb:296:in `block in start_thread'
```

**Comment by mikeng13** [comment]
Also, please merge the latest changes resulting from closing this PR: https://github.com/optimizely/ruby-sdk/pull/61.

File: lib/optimizely/project_config.rb [hunk 1]
```diff
@@ -246,19 +253,113 @@ def get_variation_id_from_key(experiment_key, variation_key)
...
+      #
+      # Returns Variation The variation which the given user and experiment should be forced into.
+
+      # check for nil and empty string user ID
+      if user_id.nil? or user_id.empty?
+        @logger.log(Logger::DEBUG, "User ID is invalid")
+        return nil
+      end
+
+      unless @forced_variation_map[user_id].present?
```
Comment (mikeng13) [inline lib/optimizely/project_config.rb]: This is causing the issue. Instead of using `.present?` we should be using `@forced_variation_map.has_key?(user_id)` 

Please replace all usages of `.present?` for object hashes with `.has_key?`

Comment (mikeng13) [reply lib/optimizely/project_config.rb]: Also we don't have tests that cover this case. Please add a regression unit test to ensure that a user ID that is not in the map will NOT crash and will return nil.

File: spec/project_config_spec.rb [hunk 2]
```diff
@@ -792,4 +792,168 @@
       expect(feature_flag).to eq(config_body['featureFlags'][0])
     end
   end
+
+  # Only those log messages have been asserted, which are directly logged in these methods.
+  # Messages that are logged in some internal function calls, are asserted in their respective function test cases. 
+  describe 'get_forced_variation' do
```
Comment (oakbani) [inline spec/project_config_spec.rb]: @mikeng13 We already have 3 invalid cases covered here. The method 'present' works on local and travis but fails on e2e testing. It's not a default method of Ruby, and is a part of active support extension. Anyway, I have replaced all of them with has_key which looks safe

File: lib/optimizely.rb [hunk 3]
```diff
@@ -131,6 +131,11 @@ def get_variation(experiment_key, user_id, attributes = nil)
         return nil
       end
 
+      unless user_id.is_a? String
```
Comment (oakbani) [inline lib/optimizely.rb]: @mikeng13 Fixing 'present' method, showed up some other apparently invalid scenarios, where a user id assigned a variation even when its null/boolean. The documentation says it has to be a string. Let me know if this is the intended behavior.
